nomad/stream/event_publisher.go

```diff
@@ -0,0 +1,220 @@
+// will receive an initial empty currentItem value which points to the first item
+// in the buffer. This allows the new subscription to call Next() without first checking
+// for the current Item.
+//
+// A Subscription will start at the requested index, or as close as possible to
+// the requested index if it is no longer in the buffer. If StartExactlyAtIndex is
+// set and the index is no longer in the buffer or not yet in the buffer an error
+// will be returned.
+//
+// When a caller is finished with the subscription it must call Subscription.Unsubscribe
+// to free ACL tracking resources. TODO(drew) ACL tracking
+func (e *EventPublisher) Subscribe(req *SubscribeRequest) (*Subscription, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	var head *bufferItem
+	var offset int
+	if req.Index != 0 {
+		head, offset = e.eventBuf.StartAtClosest(req.Index)
+	} else {
+		head = e.eventBuf.Head()
+	}
+	if offset > 0 && req.StartExactlyAtIndex {
+		return nil, fmt.Errorf("requested index not in buffer")
+	} else {
+		e.logger.Debug("requested index no longer in buffer", "requsted", int(req.Index), "closest", int(head.Events.Index))
+	}
+
+	// Empty head so that calling Next on sub
+	start := newBufferItem(&structs.Events{Index: req.Index})
+	start.link.next.Store(head)
+	close(start.link.ch)
+
+	sub := newSubscription(req, start, e.subscriptions.unsubscribe(req))
+
+	e.subscriptions.add(req, sub)
+	return sub, nil
+}
+
+// CloseAll closes all subscriptions
+func (e *EventPublisher) CloseAll() {
+	e.subscriptions.closeAll()
+}
+
+func (e *EventPublisher) handleUpdates(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			e.subscriptions.closeAll()
+			return
+		case update := <-e.publishCh:
+			e.sendEvents(update)
+		}
+	}
+}
+
+// sendEvents sends the given events to the publishers event buffer.
+func (e *EventPublisher) sendEvents(update *structs.Events) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	e.eventBuf.Append(update)
+}
+
+func (s *subscriptions) add(req *SubscribeRequest, sub *Subscription) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	subsByToken, ok := s.byToken[req.Token]
+	if !ok {
+		subsByToken = make(map[*SubscribeRequest]*Subscription)
+		s.byToken[req.Token] = subsByToken
+	}
+	subsByToken[req] = sub
+}
+
+func (s *subscriptions) closeSubscriptionsForTokens(tokenSecretIDs []string) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	for _, secretID := range tokenSecretIDs {
+		if subs, ok := s.byToken[secretID]; ok {
+			for _, sub := range subs {
+				sub.forceClose()
+			}
+		}
+	}
+}
+
+// unsubscribe returns a function that the subscription will call to remove
+// itself from the subsByToken.
+// This function is returned as a closure so that the caller doesn't need to keep
+// track of the SubscriptionRequest, and can not accidentally call unsubscribe with the
+// wrong pointer.
+func (s *subscriptions) unsubscribe(req *SubscribeRequest) func() {
```

Comment:
How do you feel about calling this `unsubscribeFn` or `unsubHandler`, given that it perform the unsubscribe the request immediately.

Reply:
sounds good to me